src/test/lib/ApiTest.cpp

```diff
@@ -2885,29 +2885,103 @@ void QuicTestTlsSetParam()
     // QUIC_PARAM_TLS_HANDSHAKE_INFO
     //
     {
-        TestScopeLogger LogScope("QUIC_PARAM_TLS_HANDSHAKE_INFO is get only");
-        QUIC_HANDSHAKE_INFO Dummy = {};
-        TEST_QUIC_STATUS(
-            QUIC_STATUS_NOT_SUPPORTED,
-            Connection.SetParam(
-                QUIC_PARAM_TLS_HANDSHAKE_INFO,
-                sizeof(Dummy),
-                &Dummy));
+        TestScopeLogger LogScope0("QUIC_PARAM_TLS_HANDSHAKE_INFO");
+        //
+        // SetParam
+        //
+        {
+            TestScopeLogger LogScope1("SetParam is not allowed");
+            QUIC_HANDSHAKE_INFO Dummy = {};
+            TEST_QUIC_STATUS(
+                QUIC_STATUS_NOT_SUPPORTED,
+                Connection.SetParam(
+                    QUIC_PARAM_TLS_HANDSHAKE_INFO,
+                    sizeof(Dummy),
+                    &Dummy));
+        }
+
+        //
+        // GetParam
+        //
+        {
+            TestScopeLogger LogScope1("GetParam");
+            uint32_t Length = 0;
+            TEST_QUIC_STATUS(
+                QUIC_STATUS_BUFFER_TOO_SMALL,
+                Connection.GetParam(
+                    QUIC_PARAM_TLS_HANDSHAKE_INFO,
+                    &Length,
+                    nullptr));
+            TEST_EQUAL(Length, sizeof(QUIC_HANDSHAKE_INFO));
+
+            //
+            // Before handshake
+            //
+            {
+                TestScopeLogger LogScope2("Before handshake");
+                QUIC_HANDSHAKE_INFO Info = {};
+                TEST_QUIC_STATUS(
```

Comment:
This is failing on schannel because of a call to `QueryContextAttributesW` (I think) not returning the expected "invalid state" error. Really, I think the product code (in TLS?) needs to explicitly check if things have been started, and if not return "invalid state" before calling into schannel or openssl. But otherwise, the test code is good. Please update the product code.

Reply:
Why do you want to change that? Just add a check before that to return invalid state?

Reply:
I _think_ you just need to check for non-NULL TLS?

Reply:
You mean this?
https://github.com/microsoft/msquic/blob/5d076de8734ec22c4b6388789b5d70cbb2f1d505/src/core/library.c#L1370-L1377

this TLS should be initialized by just ConnectionStart

Reply:
Ah, yeah. I guess it's already validated there.

Reply:
Should I add
`else if (Connection->Crypto.TlsState.HandshakeComplete) {Status = QUIC_STATUS_INVALID_STATE}`
here?
If so, all the parameters processed in CxPlatTlsParamGet depends on the HandshakeComplete.
Two common parameters looks good, but not sure other 3 parameters should be depends on it.
QUIC_PARAM_TLS_SCHANNEL_CONTEXT_ATTRIBUTE_W
QUIC_PARAM_TLS_SCHANNEL_CONTEXT_ATTRIBUTE_EX_W
QUIC_PARAM_TLS_SCHANNEL_SECURITY_CONTEXT_TOKEN

Can I treat `Only valid until the QUIC_CONNECTION_EVENT_CONNECTED event, or when TLS is cleaned up.` as after HandshakeComplete?

Reply:
Schannel case doesn't work for all 5 parameters with non-HandshakeComplete state.
The 3 Schannel specific parameters have not been tested with HandshakeComplete state though
Might need to update TlsTest.cpp